browser/brave_wallet/asset_discovery_manager_unittest.cc

```diff
@@ -0,0 +1,1190 @@
+    // DiscoverAssetsOnAllSupportedChains* call in this unit test.
+    asset_discovery_manager_->remaining_chains_ = 1;
+    asset_discovery_manager_->SetDiscoverAssetsCompletedCallbackForTesting(
+        base::BindLambdaForTesting(
+            [&](const std::string& chain_id,
+                const std::vector<mojom::BlockchainTokenPtr> discovered_assets,
+                mojom::ProviderError error, const std::string& error_message) {
+              EXPECT_EQ(chain_id, chain_id);
+              EXPECT_EQ(expected_error, error);
+              EXPECT_EQ(expected_error_message, error_message);
+              ASSERT_EQ(expected_token_contract_addresses.size(),
+                        discovered_assets.size());
+              for (size_t i = 0; i < discovered_assets.size(); i++) {
+                EXPECT_EQ(expected_token_contract_addresses[i],
+                          discovered_assets[i]->contract_address);
+              }
+            }));
+    asset_discovery_manager_->DiscoverAssets(
+        chain_id, mojom::CoinType::ETH, account_addresses, false,
+        kEthereumBlockTagEarliest, kEthereumBlockTagLatest);
+    wallet_service_observer_->WaitForOnDiscoverAssetsCompleted(
+        expected_token_contract_addresses);
+    auto next_asset_discovery_from_blocks =
+        GetPrefs()->GetDict(kBraveWalletNextAssetDiscoveryFromBlocks).Clone();
+    const auto path = base::StrCat({kEthereumPrefKey, ".", chain_id});
+    const std::string* next_asset_discovery_from_block =
+        next_asset_discovery_from_blocks.FindStringByDottedPath(path);
+    if (next_asset_discovery_from_block) {
+      EXPECT_EQ(*next_asset_discovery_from_block,
+                expected_next_asset_discovery_from_block);
+    } else {
+      ASSERT_EQ(expected_next_asset_discovery_from_block, "");
+    }
+    wallet_service_observer_->Reset();
+  }
+
+  void TestDiscoverAssetsOnAllSupportedChainsAccountsAdded(
+      const std::vector<std::string>& account_addresses,
+      const std::vector<std::string>& expected_chain_ids,
+      const base::Time expected_assets_last_discovered_at_pref,
+      const base::Value::Dict expected_next_asset_discovery_from_blocks,
+      const std::vector<std::string>& expected_token_contract_addresses = {}) {
+    asset_discovery_manager_->DiscoverAssetsOnAllSupportedChainsAccountsAdded(
+        account_addresses);
+    base::RunLoop().RunUntilIdle();
+    EXPECT_EQ(GetPrefs()->GetTime(kBraveWalletLastDiscoveredAssetsAt),
+              expected_assets_last_discovered_at_pref);
+    EXPECT_EQ(GetPrefs()->GetDict(kBraveWalletNextAssetDiscoveryFromBlocks),
+              expected_next_asset_discovery_from_blocks);
+    EXPECT_FALSE(wallet_service_observer_->OnDiscoverAssetsCompletedFired());
+    wallet_service_observer_->Reset();
+  }
+
+  void TestDiscoverAssetsOnAllSupportedChainsRefresh(
+      const std::vector<std::string>& account_addresses,
+      const base::Time previous_assets_last_discovered_at,
```

Comment:
I think it's better to just get the pref value in this function before starting discovery rather than pass the value in.